torchvision/prototype/datasets/utils/_internal.py

```diff
@@ -118,8 +118,14 @@ def getitem(*items: Any) -> Callable[[Any], Any]:
     return functools.partial(_getitem_closure, items=items)
 
 
+def _getattr_closure(obj: Any, *, attrs: Sequence[str]) -> Any:
+    for attr in attrs:
+        obj = getattr(obj, attr)
+    return obj
```

Comment:
This enables chained `getattr` calls. In turn, this allows me to use this filter function

```python
images_dp = Filter(images_dp, path_comparator("parent.name", config.split))
```

rather than writing a custom function that extracts the name of the parent folder.

Reply:
I'll leave it up to you to decide, as the entire logic of `path_comparator` and `path_accessor` is honestly too complex for me to handle ATM.

But as rule of thumb, I tend to avoid changes to helper functions when they just only address one single use-case.

Reply:
> as the entire logic of `path_comparator` and `path_accessor` is honestly too complex for me to handle ATM.

The complexity stems from the fact that we cannot use lambdas or local function together with the datapipes, since they cannot be serialized safely. Quick explanation:

- `path_accessor`: When using datapipes you often have path-handle-tuples for files. `path_accessor` takes such a tuple, turns the the first item into a `pathlib.Path` and accesses it according to the input. So `path_accessor("parent.name")` is equivalent to

    ```python
    def extract_folder_name(data):
        path = pathlib.Path(data[0])
        return path.parent.name
    ```

    This is useful when you want to merge two datapipes and need a function that generates merge keys.
- `path_comparator`: This is one layer on top of `path_accessor` by providing an equality comparison for the extract path information. For example `Filter(scenes_dp, path_comparator("name", f"CLEVR_{config.split}_scenes.json"))` will select all files where the file name matches the second attribute. (This is what I used to refactor your suggestion about the regex).

> But as rule of thumb, I tend to avoid changes to helper functions when they just only address one single use-case.

It is true, that this currently only addresses this, but I already needed to make the same changes to `getitem` a while back. So I'm guessing if I don't do it now, I'll have to do it in the future anyway, but then I also have to remember that I need to go back to here and fix this.